File: packages/web/src/lib/school.ts
import { useEffect, useState } from 'react';
import {
  arrayUnion,
  collection,
  deleteDoc,
  deleteField,
  doc,
  getDoc,
  getDocs,
  increment,
  onSnapshot,
  query,
  serverTimestamp,
  setDoc,
  updateDoc,
  where,
  writeBatch,
  type DocumentData,
  type QueryDocumentSnapshot,
} from 'firebase/firestore';
import type {
  AccessRequest,
  ClassroomGrantLevel,
  MemberRole,
  MemberScope,
  PendingInvite,
  School,
  SchoolMember,
} from '@sprout/shared';
import { firebaseClient } from './firebase';
import { assignClassroomOwner } from './firestore';

const db = firebaseClient.firestore;

function normalizeEmail(email: string): string {
  return email.trim().toLowerCase();
}

function memberFromDoc(d: QueryDocumentSnapshot<DocumentData>): SchoolMember {
  const data = d.data();
  return {
    uid: d.id,
    role: data.role,
    displayName: data.displayName ?? '',
    email: data.email ?? '',
    firstName: data.firstName,
    lastName: data.lastName,
    staffId: data.staffId,
    scope: data.scope,
    classroomGrants: data.classroomGrants,
    addedByUid: data.addedByUid,
    createdAt: data.createdAt?.toDate() ?? new Date(),
  };
}

function accessRequestFromDoc(d: QueryDocumentSnapshot<DocumentData>): AccessRequest {
  const data = d.data();
  return {
    id: d.id,
    schoolId: data.schoolId,
    contextId: data.contextId,
    contextName: data.contextName,
    requestedByUid: data.requestedByUid,
    requestedByDisplayName: data.requestedByDisplayName,
    targetUid: data.targetUid,
    targetDisplayName: data.targetDisplayName,
    level: data.level,
    status: data.status,
    createdAt: data.createdAt?.toDate() ?? new Date(),
    resolvedByUid: data.resolvedByUid,
    resolvedAt: data.resolvedAt?.toDate(),
  };
}

function inviteFromDoc(d: QueryDocumentSnapshot<DocumentData>): PendingInvite {
  const data = d.data();
  return {
    email: d.id,
    schoolId: data.schoolId,
    role: data.role,
    scope: data.scope,
    firstName: data.firstName,
    lastName: data.lastName,
    staffId: data.staffId,
    invitedByUid: data.invitedByUid,
    createdAt: data.createdAt?.toDate() ?? new Date(),
  };
}

/** The caller becomes the school's founding super_admin. Also upserts
 * `users/{founderUid}` and adds the school to `users/{founderUid}.schoolIds`
 * — mirrors createClassroom's first-write-creates-profile pattern.
 *
 * The school doc is created in its own write, awaited BEFORE the member
 * doc — not batched together. firestore.rules' isFoundingSuperAdmin reads
 * the school doc via get() to verify founderUid; within a single atomic
 * batch/transaction, get()/exists() calls see the pre-commit snapshot for
 * every write in that batch, so a member-doc write bundled into the same
 * batch as the school-doc write can never see the school as existing yet
 * — the create is unconditionally denied. Splitting them is the fix, not
 * a stylistic choice (found via real end-to-end testing, not caught by
 * rules-unit-testing, which happens to exercise these as two independently
 * awaited setDoc calls rather than a real batch). */
export async function createSchool({
  name,
  founderUid,
  founderDisplayName,
  founderEmail,
  nces,
}: {
  name: string;
  founderUid: string;
  founderDisplayName?: string | null;
  founderEmail?: string | null;
  /** Optional — set when the founder picked a match from the NCES lookup
   * rather than typing a name manually. Purely informational (displayed on
   * the school page), not consulted by rules or anything else. */
  nces?: { ncesId: string; street: string; city: string; state: string; zip: string };
}): Promise<string> {
  const schoolRef = doc(collection(db, 'schools'));
  // founderUid/superAdminCount are rules bootstrap/invariant plumbing (see
  // firestore.rules' isFoundingSuperAdmin) — not modeled on School.
  await setDoc(schoolRef, {
    name,
    founderUid,
    superAdminCount: 1,
    ...(nces ? { nces } : {}),
    createdAt: serverTimestamp(),
  });

  // These two don't depend on seeing each other mid-batch (the member
  // create only needs the already-committed school doc above; the users/
  // {uid} write is always allowed via isOwner regardless), so they're
  // safe to batch together.
  const batch = writeBatch(db);
  batch.set(doc(db, 'schools', schoolRef.id, 'members', founderUid), {
    role: 'super_admin',
    displayName: founderDisplayName ?? null,
    email: founderEmail ?? null,
    addedByUid: founderUid,
    createdAt: serverTimestamp(),
  });
  batch.set(
    doc(db, 'users', founderUid),
    {
      displayName: founderDisplayName ?? null,
      email: founderEmail ?? null,
      schoolIds: arrayUnion(schoolRef.id),
      createdAt: serverTimestamp(),
    },
    { merge: true },
  );
  await batch.commit();

  return schoolRef.id;
}

export function useSchoolIdsForUser(uid: string): string[] {
  const [schoolIds, setSchoolIds] = useState<string[]>([]);

  useEffect(() => {
    return onSnapshot(doc(db, 'users', uid), (snapshot) => {
      setSchoolIds(snapshot.data()?.schoolIds ?? []);
    });
  }, [uid]);

  return schoolIds;
}

export async function getSchool(schoolId: string): Promise<School | null> {
  const snapshot = await getDoc(doc(db, 'schools', schoolId));
  const data = snapshot.data();
  if (!data) return null;
  return {
    id: snapshot.id,
    name: data.name,
    ...(data.nces ? { nces: data.nces } : {}),
    ...(data.enabledGrades ? { enabledGrades: data.enabledGrades } : {}),
    createdAt: data.createdAt?.toDate() ?? new Date(),
  };
}

/** Name and enabledGrades only — no nces edit (no UI surface displays those
 * fields for editing today) and no delete (rules permit it, but there's no
 * cascade-delete for the school's contexts/students/members/invites). */
export async function updateSchool(
  schoolId: string,
  updates: { name?: string; enabledGrades?: string[] },
): Promise<void> {
  await updateDoc(doc(db, 'schools', schoolId), { ...updates });
}

/** Undefined while loading, null once loaded if the user isn't a member. */
export function useMyMembership(schoolId: string | undefined, uid: string): SchoolMember | null | undefined {
  const [member, setMember] = useState<SchoolMember | null | undefined>(undefined);

  useEffect(() => {
    if (!schoolId) {
      setMember(null);
      return;
    }
    return onSnapshot(doc(db, 'schools', schoolId, 'members', uid), (snapshot) => {
      setMember(snapshot.exists() ? memberFromDoc(snapshot) : null);
    });
  }, [schoolId, uid]);

  return member;
}

/** A delegate admin can remove a teacher; only a super_admin can remove an
 * admin or another super_admin (hierarchical delegation — see
 * firestore.rules). Removing the school's last super_admin is denied by the
 * rules regardless of who attempts it. Removing your own membership this
 * way is possible but has no dedicated "leave school" UI yet.
 *
 * Also auto-unassigns any classroom(s) this member owned in this school —
 * otherwise a removed teacher would still satisfy isContextOwner on those
 * classrooms and their existing students forever (ownerUids is a plain
 * array-contains check independent of school membership). Unassign runs
 * before the member delete, not batched with it: if the delete step failed
 * after a successful unassign, the worst case is a briefly-ownerless
 * classroom while the departing teacher is still technically a member (an
 * admin can just reassign it) — the reverse ordering would recreate the
 * exact security gap this exists to close. */
export async function removeMember(schoolId: string, uid: string): Promise<void> {
  const memberRef = doc(db, 'schools', schoolId, 'members', uid);
  const memberSnapshot = await getDoc(memberRef);
  const role = memberSnapshot.data()?.role;

  const ownedClassrooms = await getDocs(
    query(collection(db, 'contexts'), where('schoolId', '==', schoolId), where('ownerUids', 'array-contains', uid)),
  );
  for (const classroomDoc of ownedClassrooms.docs) {
    await assignClassroomOwner(classroomDoc.id, null);
  }

  if (role === 'super_admin') {
    // Paired with firestore.rules' superAdminCount > 1 check — kept in
    // sync here, not independently re-derived by the rules.
    const batch = writeBatch(db);
    batch.delete(memberRef);
    batch.update(doc(db, 'schools', schoolId), { superAdminCount: increment(-1) });
    await batch.commit();
  } else {
    await deleteDoc(memberRef);
  }
}

/** Changes an existing teacher's scope only — never role. Role changes
 * (promoting/demoting across teacher/admin/super_admin) stay a two-step
 * remove-then-reinvite flow; firestore.rules' members update rule only
 * permits writes that leave role as 'teacher'. */
export async function updateMemberScope(schoolId: string, uid: string, scope: MemberScope): Promise<void> {
  await updateDoc(doc(db, 'schools', schoolId, 'members', uid), { scope });
}

export function useMembersOfSchool(schoolId: string | undefined): SchoolMember[] {
  const [members, setMembers] = useState<SchoolMember[]>([]);

  useEffect(() => {
    if (!schoolId) {
      setMembers([]);
      return;
    }
    return onSnapshot(collection(db, 'schools', schoolId, 'members'), (snapshot) => {
      setMembers(snapshot.docs.map(memberFromDoc));
    });
  }, [schoolId]);

  return members;
}

/** Records who's coming, keyed by email — not tied to an existing account.
 * Activated automatically the first time that email signs in (see
 * claimPendingInviteIfAny), whether that's before or after this call. */
export async function inviteMember({
  schoolId,
  email,
  role,
  scope,
  firstName,
  lastName,
  staffId,
  invitedByUid,
}: {
  schoolId: string;
  email: string;
  role: MemberRole;
  scope?: MemberScope;
  firstName?: string;
  lastName?: string;
  staffId?: string;
  invitedByUid: string;
}): Promise<void> {
  await writeBatch(db)
    .set(doc(db, 'pendingInvites', normalizeEmail(email)), {
      schoolId,
      role,
      ...(scope ? { scope } : {}),
      ...(firstName ? { firstName } : {}),
      ...(lastName ? { lastName } : {}),
      ...(staffId ? { staffId } : {}),
      invitedByUid,
      createdAt: serverTimestamp(),
    })
    .commit();
}

export function usePendingInvitesForSchool(schoolId: string | undefined): PendingInvite[] {
  const [invites, setInvites] = useState<PendingInvite[]>([]);

  useEffect(() => {
    if (!schoolId) {
      setInvites([]);
      return;
    }
    const q = query(collection(db, 'pendingInvites'), where('schoolId', '==', schoolId));
    return onSnapshot(q, (snapshot) => setInvites(snapshot.docs.map(inviteFromDoc)));
  }, [schoolId]);

  return invites;
}

export async function cancelInvite(email: string): Promise<void> {
  await deleteDoc(doc(db, 'pendingInvites', normalizeEmail(email)));
}

/** Runs once after every sign-in, on every platform: if a pending invite
 * exists for `email`, activates the access it configured (creates the
 * members doc, appends to users/{uid}.schoolIds) and deletes the invite.
 * No-op if there's no matching invite — the normal case for every
 * self-serve/standalone user. */
export async function claimPendingInviteIfAny({
  uid,
  email,
  displayName,
}: {
  uid: string;
  email: string;
  displayName?: string | null;
}): Promise<void> {
  const normalized = normalizeEmail(email);
  const inviteRef = doc(db, 'pendingInvites', normalized);
  const inviteSnapshot = await getDoc(inviteRef);
  const invite = inviteSnapshot.data();
  if (!invite) return;

  const batch = writeBatch(db);
  batch.set(doc(db, 'schools', invite.schoolId, 'members', uid), {
    role: invite.role,
    displayName: displayName ?? null,
    email: normalized,
    ...(invite.scope ? { scope: invite.scope } : {}),
    ...(invite.firstName ? { firstName: invite.firstName } : {}),
    ...(invite.lastName ? { lastName: invite.lastName } : {}),
    ...(invite.staffId ? { staffId: invite.staffId } : {}),
    addedByUid: uid,
    createdAt: serverTimestamp(),
  });
  if (invite.role === 'super_admin') {
    batch.update(doc(db, 'schools', invite.schoolId), { superAdminCount: increment(1) });
  }
  batch.set(doc(db, 'users', uid), { schoolIds: arrayUnion(invite.schoolId) }, { merge: true });
  batch.delete(inviteRef);
  await batch.commit();
}

/** A classroom owner proposing that a colleague (an existing active
 * teacher member of the school) get 'award' or 'manage' access to
 * specifically their classroom. Grants nothing by itself — only an
 * admin/super_admin approving it (see approveAccessRequest) actually
 * grants access, keeping "only admins/super_admins grant access" intact. */
export async function createAccessRequest({
  schoolId,
  contextId,
  contextName,
  requestedByUid,
  requestedByDisplayName,
  targetUid,
  targetDisplayName,
  level,
}: {
  schoolId: string;
  contextId: string;
  contextName: string;
  requestedByUid: string;
  requestedByDisplayName: string;
  targetUid: string;
  targetDisplayName: string;
  level: ClassroomGrantLevel;
}): Promise<void> {
  await writeBatch(db)
    .set(doc(collection(db, 'accessRequests')), {
      schoolId,
      contextId,
      contextName,
      requestedByUid,
      requestedByDisplayName,
      targetUid,
      targetDisplayName,
      level,
      status: 'pending',
      createdAt: serverTimestamp(),
    })
    .commit();
}

export function usePendingAccessRequestsForSchool(schoolId: string | undefined): AccessRequest[] {
  const [requests, setRequests] = useState<AccessRequest[]>([]);

  useEffect(() => {
    if (!schoolId) {
      setRequests([]);
      return;
    }
    const q = query(
      collection(db, 'accessRequests'),
      where('schoolId', '==', schoolId),
      where('status', '==', 'pending'),
    );
    return onSnapshot(q, (snapshot) => setRequests(snapshot.docs.map(accessRequestFromDoc)));
  }, [schoolId]);

  return requests;
}

/** A classroom's own request history (any status) — for the owner's view
 * on the classroom detail page. */
export function useAccessRequestsForContext(contextId: string | undefined): AccessRequest[] {
  const [requests, setRequests] = useState<AccessRequest[]>([]);

  useEffect(() => {
    if (!contextId) {
      setRequests([]);
      return;
    }
    const q = query(collection(db, 'accessRequests'), where('contextId', '==', contextId));
    return onSnapshot(q, (snapshot) => setRequests(snapshot.docs.map(accessRequestFromDoc)));
  }, [contextId]);

  return requests;
}

/** Approving is the only thing that actually grants access: updates the
 * request's status and, in the same batch, writes the requested level
 * onto the target's classroomGrants — mirrors updateMemberScope's shape. */
export async function approveAccessRequest(request: AccessRequest, resolvedByUid: string): Promise<void> {
  const batch = writeBatch(db);
  batch.update(doc(db, 'accessRequests', request.id), {
    status: 'approved',
    resolvedByUid,
    resolvedAt: serverTimestamp(),
  });
  batch.update(doc(db, 'schools', request.schoolId, 'members', request.targetUid), {
    [`classroomGrants.${request.contextId}`]: request.level,
  });
  await batch.commit();
}

export async function declineAccessRequest(requestId: string, resolvedByUid: string): Promise<void> {
  await updateDoc(doc(db, 'accessRequests', requestId), {
    status: 'declined',
    resolvedByUid,
    resolvedAt: serverTimestamp(),
  });
}

/** The requester cancelling their own still-pending request. */
export async function cancelAccessRequest(requestId: string): Promise<void> {
  await deleteDoc(doc(db, 'accessRequests', requestId));
}

/** Admin-direct revoke — grants shouldn't be permanent with no way back
 * short of editing Firestore. */
export async function revokeClassroomGrant(schoolId: string, uid: string, contextId: string): Promise<void> {
  await updateDoc(doc(db, 'schools', schoolId, 'members', uid), {
    [`classroomGrants.${contextId}`]: deleteField(),
  });
}

File: packages/mobile/lib/core/models/school.dart
/// A school in the security matrix (BR-1.3.11/1.3.12). See firestore.rules
/// for the enforcement side of this model. `founderUid`/`superAdminCount`
/// are rules bootstrap/invariant plumbing, not app-facing — deliberately
/// not modeled here; [SchoolRepository] writes them directly.
class School {
  const School({required this.id, required this.name});

  final String id;
  final String name;
}

/// Hierarchical delegation: only a super_admin can create/remove another
/// super_admin or an admin. A school is never left without at least one
/// super_admin (enforced in firestore.rules, not just the UI).
enum MemberRole { superAdmin, admin, teacher }

enum MemberScopeType { own, grades, school }

/// Meaningless for admins (implicit whole-school access). For teachers:
/// [MemberScopeType.own] = only classrooms they directly own (the default);
/// [MemberScopeType.grades] = any classroom in the school whose gradeLevel
/// is in [grades]; [MemberScopeType.school] = every classroom in the school
/// (the PE/art/music case).
class MemberScope {
  const MemberScope.own() : type = MemberScopeType.own, grades = const [];
  const MemberScope.grades(this.grades) : type = MemberScopeType.grades;
  const MemberScope.school() : type = MemberScopeType.school, grades = const [];
  const MemberScope({required this.type, this.grades = const []});

  final MemberScopeType type;
  final List<String> grades;

  Map<String, dynamic> toJson() => {
        'type': type.name,
        if (type == MemberScopeType.grades) 'grades': grades,
      };

  static MemberScope fromJson(Map<String, dynamic> json) {
    final typeName = json['type'] as String;
    switch (typeName) {
      case 'grades':
        return MemberScope.grades(List<String>.from(json['grades'] as List? ?? const []));
      case 'school':
        return const MemberScope.school();
      default:
        return const MemberScope.own();
    }
  }
}

/// 'award' = record earn/spend transactions only; 'manage' = full
/// rename/delete/roster rights, same as the classroom's owner. Only ever
/// set via an admin approving an [AccessRequest] (or an admin editing it
/// directly) — an owner can only propose, never grant directly.
enum ClassroomGrantLevel { award, manage }

ClassroomGrantLevel _grantLevelFromJson(String value) =>
    value == 'manage' ? ClassroomGrantLevel.manage : ClassroomGrantLevel.award;

class SchoolMember {
  const SchoolMember({
    required this.uid,
    required this.role,
    required this.displayName,
    required this.email,
    this.firstName,
    this.lastName,
    this.staffId,
    this.scope,
    this.classroomGrants = const {},
  });

  final String uid;
  final MemberRole role;
  final String displayName;
  final String email;

  /// Pure supplementary roster metadata (manual edit or CSV import) —
  /// never overwrites displayName, never consulted by the invite/claim
  /// identity logic. displayName (auth-provider-sourced) stays the one
  /// authoritative name.
  final String? firstName;
  final String? lastName;
  final String? staffId;

  final MemberScope? scope;

  /// contextId -> grant level, for classrooms this member doesn't own and
  /// whose grade/scope wouldn't otherwise cover.
  final Map<String, ClassroomGrantLevel> classroomGrants;

  SchoolMember copyWith({MemberScope? scope, Map<String, ClassroomGrantLevel>? classroomGrants}) =>
      SchoolMember(
        uid: uid,
        role: role,
        displayName: displayName,
        email: email,
        firstName: firstName,
        lastName: lastName,
        staffId: staffId,
        scope: scope ?? this.scope,
        classroomGrants: classroomGrants ?? this.classroomGrants,
      );

  static Map<String, ClassroomGrantLevel> classroomGrantsFromJson(Object? value) {
    if (value is! Map) return const {};
    return value.map((key, v) => MapEntry(key as String, _grantLevelFromJson(v as String)));
  }
}

/// Doc ID is the invitee's lowercased email. Claimed automatically the
/// first time that email signs in, on whichever platform they use.
class PendingInvite {
  const PendingInvite({
    required this.email,
    required this.schoolId,
    required this.role,
    this.scope,
    this.firstName,
    this.lastName,
    this.staffId,
  });

  final String email;
  final String schoolId;
  final MemberRole role;
  final MemberScope? scope;
  final String? firstName;
  final String? lastName;
  final String? staffId;
}

/// Doc ID is the target student's real school email, lowercased. Created
/// by staff with manage access to the named student's classroom; claimed
/// automatically the first time that email signs in — mirrors
/// [PendingInvite], except the claim writes onto an EXISTING
/// students/{studentId} doc rather than creating a new doc under the
/// claimant's own uid. Deliberately carries no contextId/schoolId — those
/// are derived from studentId itself so there's no spoofable field for a
/// caller to lie about which classroom they manage.
class PendingStudentLink {
  const PendingStudentLink({
    required this.email,
    required this.studentId,
    required this.invitedByUid,
  });

  final String email;
  final String studentId;
  final String invitedByUid;
}

enum AccessRequestStatus { pending, approved, declined }

AccessRequestStatus _accessRequestStatusFromJson(String value) {
  switch (value) {
    case 'approved':
      return AccessRequestStatus.approved;
    case 'declined':
      return AccessRequestStatus.declined;
    default:
      return AccessRequestStatus.pending;
  }
}

/// A classroom owner's request that a colleague (an existing active
/// teacher member of the school) get 'award' or 'manage' access to
/// specifically their classroom — fulfilled only by an admin/super_admin,
/// who writes the resulting grant onto the target's classroomGrants. Keeps
/// "only admins/super_admins grant access" intact: an owner can only
/// propose.
class AccessRequest {
  const AccessRequest({
    required this.id,
    required this.schoolId,
    required this.contextId,
    required this.contextName,
    required this.requestedByUid,
    required this.requestedByDisplayName,
    required this.targetUid,
    required this.targetDisplayName,
    required this.level,
    required this.status,
    this.resolvedByUid,
  });

  final String id;
  final String schoolId;
  final String contextId;
  final String contextName;
  final String requestedByUid;
  final String requestedByDisplayName;
  final String targetUid;
  final String targetDisplayName;
  final ClassroomGrantLevel level;
  final AccessRequestStatus status;
  final String? resolvedByUid;

  AccessRequest copyWith({AccessRequestStatus? status, String? resolvedByUid}) => AccessRequest(
        id: id,
        schoolId: schoolId,
        contextId: contextId,
        contextName: contextName,
        requestedByUid: requestedByUid,
        requestedByDisplayName: requestedByDisplayName,
        targetUid: targetUid,
        targetDisplayName: targetDisplayName,
        level: level,
        status: status ?? this.status,
        resolvedByUid: resolvedByUid ?? this.resolvedByUid,
      );

  static AccessRequestStatus statusFromJson(String value) => _accessRequestStatusFromJson(value);
  static ClassroomGrantLevel levelFromJson(String value) => _grantLevelFromJson(value);
}
